### PR TITLE
Add initial code base for a common repository for coding standards

### PR DESCRIPTION
**Pull request by AltamashShaikh**
### Description:

Add initial code base for a common repository for coding standards

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: README.md
Comment (michalkleiner) [inline README.md:1]: Do we need to have the paths in the instructions?
Normally we would include this repository as a dev dependency in projects/plugins composer.json and then have composer script e.g. `codecheck` (which you'd run via `composer codecheck` and all the necessary paths would be either default or defined in the composer script params.

So we should suggest, if anything, how to add a composer script line for a given plugin. Most of the times I don't expect we'd need any specific config as all composer scripts run primarily in the folder where composer.json file is located.

File: Standard/GoogleAnalyticsImporter/phpcs.xml
Comment (michalkleiner) [inline Standard/GoogleAnalyticsImporter/phpcs.xml:1]: if this is the default phpcs.xml, it should be in the folder. We don't need specific plugin names or folders in the path to the default phpcs.xml

Comment (michalkleiner) [reply Standard/GoogleAnalyticsImporter/phpcs.xml:1]: we will have a default here, one that covers as much as possible, and then any customisations would be made in module's (plugin's) repo.

File: composer.json
Comment (michalkleiner) [inline composer.json]:
```suggestion
    "name": "matomo-org/matomo-coding-standards",
```

File: Standard/GoogleAnalyticsImporter/phpcs.xml
Comment (AltamashShaikh) [reply Standard/GoogleAnalyticsImporter/phpcs.xml:1]: @michalkleiner Its just a skeleton to decide how we want this repository to be.

File: README.md
Comment (AltamashShaikh) [reply README.md:1]: I just added that, so that its easier for reviewer to run and check, we can tweak this.

File: Standard/GoogleAnalyticsImporter/phpcs.xml
Comment (sgiehl) [inline Standard/GoogleAnalyticsImporter/phpcs.xml:1]: As this should be a global ruleset the file should be named `Matomo/ruleset.xml`.

File: Standard/GoogleAnalyticsImporter/phpcs.xml
Comment (sgiehl) [inline Standard/GoogleAnalyticsImporter/phpcs.xml]:
```suggestion
    <description>Matomo Coding Standard</description>
```

File: Standard/GoogleAnalyticsImporter/phpcs.xml
Comment (sgiehl) [inline Standard/GoogleAnalyticsImporter/phpcs.xml]: In the default ruleset we should only exclude vendor and maybe lib. If core or a certain plugin needs additional exclude, they need to be defined in the according repo

```suggestion
    <exclude-pattern>*/vendor/*</exclude-pattern>
    <exclude-pattern>*/libs/*</exclude-pattern>
```

File: Standard/GoogleAnalyticsImporter/phpcs.xml
Comment (sgiehl) [inline Standard/GoogleAnalyticsImporter/phpcs.xml]: we have meanwhile removed those in core. 

```suggestion
```

File: Standard/GoogleAnalyticsImporter/phpcs.xml
Comment (sgiehl) [inline Standard/GoogleAnalyticsImporter/phpcs.xml]: To make that working in core and plugin repos we should change that to
```suggestion
        <exclude-pattern>*/Updates/*</exclude-pattern>
```

File: Standard/GoogleAnalyticsImporter/phpcs.xml
Comment (sgiehl) [inline Standard/GoogleAnalyticsImporter/phpcs.xml]: As those rules should be activated at some point, we should remove those from the global ruleset and add it to the ones in each repo, as long as needed.

File: composer.json
Comment (sgiehl) [inline composer.json]: Those should be moved to `require` instead of `require-dev`

**Review by sgiehl — CHANGES_REQUESTED** [review]
Left a couple of comments with things that imho should be changed in global phpcs ruleset.
Once this global ruleset is finished, we can start using it in repo local phpcs files.

**Comment by AltamashShaikh** [comment]
> Left a couple of comments with things that imho should be changed in global phpcs ruleset. Once this global ruleset is finished, we can start using it in repo local phpcs files.

@sgiehl I copied the one I created for GAImporter plugin, I should copy the one we have in [core](https://github.com/matomo-org/matomo/blob/5.x-dev/phpcs.xml) right ?

**Comment by sgiehl** [comment]
No. It needs to be different from core one as well

**Comment by AltamashShaikh** [comment]
@sgiehl @michalkleiner Does this look good to you guys now ??

**Comment by sgiehl** [comment]
@AltamashShaikh I would actually remove the plugin specific file here. The repo should only contain the base ruleset we want to use. In addition we can consider to add custom sniffs at some point when needed.
Every additional stuff for adjusting used rules or excluding additional files should then be handled by `phpcs.xml` files in the according repo. E.g. for core that could then look like this: https://github.com/matomo-org/matomo/compare/use_matomo_cs

**Comment by AltamashShaikh** [comment]
@sgiehl Removed plugin specific coding standard, we need to make this repository public before merging in core, do you have access to do the needful ?

**Comment by sgiehl** [comment]
@AltamashShaikh No I don't have access to do that.

**Comment by AltamashShaikh** [comment]
@caddoo Can you make this repository public ?

**Comment by michalkleiner** [comment]
We should also change the default branch to `main` (inclusive naming) and when this is merged tag it as `0.0.1` or similar so that we use a version such as `~0.0.1`, not a branch.

File: Matomo/ruleset.xml
Comment (michalkleiner) [inline Matomo/ruleset.xml]:
```suggestion
    <!-- Use PSR-12 standard as default -->
```

File: composer.json
Comment (michalkleiner) [inline composer.json]: This repo will grow to also include front-end CS and other rules, so we should describe the package in a generic way.
```suggestion
    "description": "Coding standards for Matomo",
```

Comment (AltamashShaikh) [reply composer.json]: @michalkleiner Makes sense :+1: , updated as suggested

**Comment by AltamashShaikh** [comment]
@tsteur Will you be able to make this repository public ?

**Comment by tsteur** [comment]
That's done @AltamashShaikh

**Comment by AltamashShaikh** [comment]
@sgiehl Can we merge this PR now and start using in core and extending the same for all the plugins ?

**Comment by sgiehl** [comment]
@AltamashShaikh I'll merge this now and see if it works as expected when used in core.

**Pull request merged**
